python/ql/src/semmle/python/dataflow/Implementation.qll

```diff
@@ -0,0 +1,849 @@
+        exists(DataFlow::Node srcnode, CallNode call, TaintKind srckind, string name |
+            src = TTaintTrackingNode_(srcnode, context, path, srckind, this) and
+            call.getFunction().(AttrNode).getObject(name) = src.getNode().asCfgNode() and
+            kind = srckind.getTaintOfMethodResult(name) and
+            node.asCfgNode() = call
+        ) and edgeLabel = "call"
+    }
+
+    pragma [noinline]
+    predicate iterationStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        exists(ForNode for, DataFlow::Node sequence, TaintKind seqkind |
+            src = TTaintTrackingNode_(sequence, context, path, seqkind, this) and
+            for.iterates(_, sequence.asCfgNode()) and
+            node.asCfgNode() = for and
+            path.noAttribute() and
+            kind = seqkind.getTaintForIteration()
+        ) and edgeLabel = "iteration"
+    }
+
+    pragma [noinline]
+    predicate parameterStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        exists(CallNode call, PythonFunctionObjectInternal pyfunc, int arg |
+            this.callWithTaintedArgument(src, call, _, pyfunc, arg, path, kind) and
+            node.asCfgNode() = pyfunc.getParameter(arg) and
+            context = TParamContext(kind, path, arg)
+        ) and edgeLabel = "parameter"
+    }
+
+    pragma [noinline]
+    predicate yieldStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        exists(DataFlow::Node srcnode, TaintKind itemkind |
+            src = TTaintTrackingNode_(srcnode, context, path, itemkind, this) and
+            itemkind = kind.getTaintForIteration() and
+            exists(PyFunctionObject func |
+                func.getFunction().isGenerator() and
+                func.getACall() = node.asCfgNode() and
+                exists(Yield yield |
+                    yield.getScope() = func.getFunction() and
+                    yield.getValue() = srcnode.asCfgNode().getNode()
+                )
+            )
+        ) and edgeLabel = "yield"
+    }
+
+    pragma [noinline]
+    predicate ifExpStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        exists(DataFlow::Node srcnode |
+            src = TTaintTrackingNode_(srcnode, context, path, kind, this) and
+            srcnode.asCfgNode() = node.asCfgNode().(IfExprNode).getAnOperand()
+        ) and edgeLabel = "if exp"
+    }
+
+    pragma [noinline]
+    predicate essaFlowStep(TaintTrackingNode src, DataFlow::Node node, TaintTrackingContext context, AttributePath path, TaintKind kind, string edgeLabel) {
+        this.(EssaTaintTracking).taintedDefinition(src, node.asVariable().getDefinition(), context, path, kind) and edgeLabel = ""
```

Comment:
Why an empty `edgeLabel`?

Reply:
There is no meaningful label here, so an empty string seemed the least misleading.
We also use it as a marker that the edge is "uninteresting" and should be merged into an "interesting" edge in the `edges` relation.